### PR TITLE
Sm/max-api-version-on-deploys

### PR DESCRIPTION
**Pull request by mshanemc**
### What does this PR do?

1. when the componentSet has made it to the SDR deploy step without any apiVersion info, set it from what's on the connection so it doesn't default to highest per mdapi coverage in the package.xml being sent
2. getConnection will use org's max, emiting a warning if the user is trying to use something too high, in case the default isn't available yet to avoid issues on other endpoints polling/cancel/etc

[NB the proper way to do this is to set `sourceApiVersion` in your `sfdx-project.json`.  This is the workaround for tooling to use more intelligent default behavior.

### What issues does this PR fix or reference?

@W-11549415@
https://github.com/forcedotcom/cli/issues/1656

### Functionality Before

```
=== Component Failures [1]

 Type  Name        Problem                        
 ───── ─────────── ────────────────────────────── 
 Error package.xml Invalid version specified:56.0 
```

### Functionality After

pushes as expected

### CODE REVIEW

File: src/client/metadataApiDeploy.ts [hunk 1]
```diff
@@ -352,6 +352,11 @@ export class MetadataApiDeploy extends MetadataTransfer<MetadataApiDeployStatus,
     const connection = await this.getConnection();
     // store for use in the scopedPostDeploy event
     this.orgId = connection.getAuthInfoFields().orgId;
+    if (this.components && !this.components.apiVersion && !this.components.sourceApiVersion) {
```
Comment (peternhale) [inline src/client/metadataApiDeploy.ts:355]: @mshanemc I am not too sure about this assertion, but looking at component set class there is a function `getObject` that [establishes apiVersion](https://github.com/forcedotcom/source-deploy-retrieve/blob/sm/max-api-version-on-deploys/src/collections/componentSet.ts#L327) for the component. Is it possible that `getObject` has already been called before reaching this check?

**Review by peternhale — CHANGES_REQUESTED** [review]

File: src/client/metadataApiDeploy.ts (hunk 1 above)
Comment (mshanemc) [reply src/client/metadataApiDeploy.ts:355]: Within SDR, that's called during the creation of the package.xml, which is pretty late.

Technically, some consumer could do anything they want with it, but it's unlikely.

Comment (mshanemc) [reply src/client/metadataApiDeploy.ts:355]: what change are you requesting?

Comment (peternhale) [reply src/client/metadataApiDeploy.ts:355]: Nothing to change if my suspicion was not confirmed. Thanks

**Review by peternhale — APPROVED** [review]

File: src/client/metadataTransfer.ts [hunk 2]
```diff
@@ -269,3 +269,16 @@ export abstract class MetadataTransfer<Status extends MetadataRequestStatus, Res
   protected abstract pre(): Promise<AsyncResult>;
   protected abstract post(result: Status): Promise<Result>;
 }
+
+/* prevent requests on apiVersions higher than the org supports */
+const getConnectionNoHigherThanOrgAllows = async (conn: Connection, requestedVersion: string): Promise<Connection> => {
```
Comment (gbockus-sf) [inline src/client/metadataTransfer.ts:274]: Do we have test coverage for this method?

File: src/client/metadataApiDeploy.ts [hunk 3]
```diff
@@ -352,6 +352,11 @@ export class MetadataApiDeploy extends MetadataTransfer<MetadataApiDeployStatus,
...
+      // if we have a ComponentSet, but got no version info, let's use the org's max version for calculating what goes into the package.xml
```
Comment (gbockus-sf) [inline src/client/metadataApiDeploy.ts:356]: thanks for this comment. def adds value.

File: src/client/metadataTransfer.ts [hunk 4]
```diff
@@ -269,3 +269,16 @@ export abstract class MetadataTransfer<Status extends MetadataRequestStatus, Res
...
+  // uses a TTL cache, so mostly won't hit the server
+  const maxApiVersion = await conn.retrieveMaxApiVersion();
+  if (requestedVersion && parseInt(requestedVersion, 10) > parseInt(maxApiVersion, 10)) {
```
Comment (gbockus-sf) [inline src/client/metadataTransfer.ts:277]: Would it be worthwhile to validate the requested version here?  It seems the caller could send an invalid version string here and get by without altering the max version allowed on the connection.

**Review by gbockus-sf — APPROVED** [review]
a few question in there but change looks good.

**Review by shetzel — APPROVED** [review]

File: src/client/metadataTransfer.ts (hunk 2 above)
Comment (mshanemc) [reply src/client/metadataTransfer.ts:274]: it's being executed on every deploy, so about 100x.
At least some of those specify an apiVersion...not sure if any don't and would check the if conditional.

File: src/client/metadataTransfer.ts (hunk 4 above)
Comment (mshanemc) [reply src/client/metadataTransfer.ts:277]: the ComponentSet allows consumers to set its apiVersion and sourceApiVersion properties.  If there's a place to put validation, it'd be there.  This is internal non-exported function, only called from this file, which is extended by Deploy and Retrieve.

Are you wanting to put getters/setters so we can do validation on those props for ComponentSet?

**Pull request merged**